src/signature/spki.rs

```diff
@@ -0,0 +1,197 @@
+/// DER ASN.1 of the key. This method is suitable when the public key is in
+/// SubjectPublicKeyInfo form (https://tools.ietf.org/html/rfc5280#section-4.1),
+/// which is a wrapper containing both the key and an identifier.
+///
+/// A common situation where this encoding is encountered is when using public keys
+/// exported by OpenSSL. If you export an RSA or ECDSA public key from a keypair
+/// with `-pubout` and friends, you will get a DER of an SPKI wrapper containing
+/// a public key. You could extract the bitstring of the key itself and use
+/// `signature:;verify`, but that is often inconvient.
+pub fn verify(signature_alg: &Algorithm,
+              public_key_spki: untrusted::Input,
+              msg: untrusted::Input,
+              signature: untrusted::Input) -> Result<(), VerifyWithSPKIError> {
+    let spki = try!(parse_spki_value(public_key_spki));
+    if !signature_alg.public_key_alg_id
+        .matches_algorithm_id_value(spki.algorithm_id_value) {
+        return Err(VerifyWithSPKIError::UnsupportedSignatureAlgorithmForPublicKey);
+    }
+    signature::verify(signature_alg.verification_alg, spki.key_value, msg,
+                      signature)
+        .map_err(|_| VerifyWithSPKIError::InvalidSignatureForPublicKey)
+}
+
+struct SubjectPublicKeyInfo<'a> {
+    algorithm_id_value: untrusted::Input<'a>,
+    key_value: untrusted::Input<'a>,
+}
+
+// Parse the public key into an algorithm OID, an optional curve OID, and the
+// key value. The caller needs to check whether these match the
+// `PublicKeyAlgorithm` for the `SignatureAlgorithm` that is matched when
+// parsing the signature.
+fn parse_spki_value(input: untrusted::Input)
+                    -> Result<SubjectPublicKeyInfo, VerifyWithSPKIError> {
+    input.read_all(VerifyWithSPKIError::BadDER, |input| {
+        let algorithm_id_value =
+        try!(der::expect_tag_and_get_value(input, der::Tag::Sequence)
+            .map_err(|_| VerifyWithSPKIError::BadDER));
+        let key_value = try!(der::bit_string_with_no_unused_bits(input)
+            .map_err(|_| VerifyWithSPKIError::BadDER));
+        Ok(SubjectPublicKeyInfo {
+            algorithm_id_value: algorithm_id_value,
+            key_value: key_value,
+        })
+    })
+}
+
+/// A signature algorithm for use when validating a signature with an SPKI-formatted public key.
+pub struct Algorithm {
+    /// The `algorithm` member in SPKI from https://tools.ietf.org/html/rfc5280#section-4.1.
+    public_key_alg_id: AlgorithmIdentifier,
+    verification_alg: &'static signature::VerificationAlgorithm,
+}
+
+// RFC 5758 Section 3.2 (ECDSA with SHA-2), and RFC 3279 Section 2.2.3 (ECDSA
+// with SHA-1) say that parameters must be omitted. RFC 4055 Section 5 and RFC
+// 3279 Section 2.2.1 both say that parameters for RSA must be encoded as NULL;
+// we relax that requirement by allowing the NULL to be omitted, to match all
+// the other signature algorithms we support and for compatibility.
```

Comment:
I believe that comment is referring to *signature* algorithm identifiers, not *key* algorithm identifiers, so it won't apply in any case since we remove `signature_alg_id` in a later commit. And yes, it appears to be outdated. I just removed it in webpki: https://github.com/briansmith/webpki/commit/2419bb730d0c8d2f315a4a38eea44d601428272d and it should be removed here as well.